### PR TITLE
Refactoring of policy and credentials

### PR DESCRIPTION
**Pull request by pingles**
* Rename CredentialsIdentity to RoleIdentity
* Remove arnResolver use within the credentials cache (it can work on the basis all credentials it's being used to manage have already got ARNs resolved)
* Policy interface changed to take role string and *Pod to avoid needing a PodGetter interface

@stefansedich hope you're ok with the rename of Credentials to Role? My reasoning was that the identity is largely used to refer to the role and not always when we have credentials.

### CODE REVIEW

**Comment by pingles** [comment]
This is also to help simplify things a little ahead of trying to incorporate #326 and #327.

**Comment by stefansedich** [comment]
> * Rename CredentialsIdentity to RoleIdentity
> * Remove arnResolver use within the credentials cache (it can work on the basis all credentials it's being used to manage have already got ARNs resolved)
> * Policy interface changed to take role string and *Pod to avoid needing a PodGetter interface
> 
> @stefansedich hope you're ok with the rename of Credentials to Role? My reasoning was that the identity is largely used to refer to the role and not always when we have credentials.

I like it! `RoleIdentity` makes much more sense. Will this one be going in soon? I will rebase on this and then continue with adding the final tests to #430 so we can get that merged.

**Comment by pingles** [comment]
Yeah hopefully. I keep making the refactorings and then changing my mind but should be in the next few days hopefully. Perhaps by the end of the week?

**Comment by stefansedich** [comment]
> Yeah hopefully. I keep making the refactorings and then changing my mind but should be in the next few days hopefully. Perhaps by the end of the week?

Sounds good! I might hold off on getting to testing then incase more changes to avoid any merge pain, but once done will get the rest of my stuff done ASAP.

**Comment by pingles** [comment]
I'll try and speed up then!

**Comment by pingles** [comment]
I've done some refactoring and merged a previous PR #327 changes.

I want to write out a truth table to confirm and give examples but this will potentially be a breaking change. Because policy is now applied after the full ARN has been resolved the regexes will resolve differently (in both directions potentially). Regexes that denied pods previously may accidentally permit them, and regexes that previously denied may now permit. I suspect that the chances are low based on the way people would likely write them but before we merge I'd like to call out examples of where this behaviour would change.

I think this is still the right thing to do, but it will mean calling it out for cluster operators in the `docs/UPGRADING.md`.

@leosunmo what do you think? Could you prep some examples of how the namespace regex behaved before, and now, and what the implications are please?

I've not refactored to change policy so that it only uses resolved ARNs (rather than calling out to the resolver) but for now it's not essential.

**Comment by stefansedich** [comment]
> @stefansedich this feels closer too; I changed code so that the `ARNResolver` converts from a role (either short-form or ARN) to a `RoleIdentity` type (which then has both `ARN` and `Role` fields).
> 
> What do you think? Next challenge is to add some more valuable tests to get codecov bot off my back :-) I've been meaning to write some better integration tests so may do that.

One thing that I am not sure off hand is in my PR for session-name and external-id I use `RoleIdentity` to add those additional bits of info, I think it should be OK I will likely just have to set those properties at a higher level after the call to `Resolve()`.

**Comment by pingles** [comment]
Ah good point. Well I'll try and get it merged asap and then it's easier to integrate and see your work and ideas.

**Comment by stefansedich** [comment]
> Ah good point. Well I'll try and get it merged asap and then it's easier to integrate and see your work and ideas.

Good stuff!

**Review by Joseph-Irving — APPROVED** [review]
I think this looks good, it would be good to understand what kind of breaking changes this may cause in relation to the regex so that we can call it out with big warning lights in the 4.0 changelog

**Comment by pingles** [comment]
> it would be good to understand what kind of breaking changes this may cause in relation to the regex so that we can call it out with big warning lights in the 4.0 

Yep, totally agree! Will add an issue for it

**Pull request merged**
